File: week4.2/sub_array_sum.py
class Solution:
    def subarraySum(self,nums:list,k:int)->int:
        # dictionary to store prefix sum count
        prefix_sum_count = {0:1}
        current_sum = 0
        count = 0 

        # loop through the array
        for num in nums:
            current_sum += num
            # check if current_sum - k is in the dictionary
            if current_sum -k in prefix_sum_count:
                count += prefix_sum_count[current_sum-k]
            # update the frequency of current prefix sum in the dictionary
            if current_sum in prefix_sum_count:
                prefix_sum_count[current_sum] += 1
            else:
                prefix_sum_count[current_sum] = 1
        return count 


s = Solution()
print(s.subarraySum(nums = [1,1,1], k = 2))

File: week4.2/sub_array_sum.js
var subArraySum =function(nums,k){
    const obj = {0:1}
    let count = 0
    let sum =0
    for(let num of nums){
        sum += num
        const diff = sum - k
        if(obj[diff]){
            count += obj[diff]
        }
        if(obj[sum]){
            obj[sum] += 1
        }else{
            obj[sum] = 1}
    }
    return count
};
console.log(subArraySum([1,1,1],2))
console.log(subArraySum([1,2,3],3))